app/services/resolver_cost_calculator.rb

```diff
@@ -0,0 +1,35 @@
+class ResolverCostCalculator
+  def initialize(source)
+    @source = source
+  end
+
+  def cost
+    @source.outcome == 'none' ? 0 : incurred_cost(@source)
+  end
+
+  private
+
+  def incurred_cost(source)
+    if source.is_a?(PartPayment)
+      if source.application.evidence_check.present?
+        fee(source) - amount_to_pay(source.application.evidence_check)
+      else
+        fee(source) - amount_to_pay(source.application)
+      end
+    else
+      fee(source)
+    end
+  end
```

Comment:
Maybe instead of `incurred_cost` as above with all the nested logic implement something like this:

```ruby
def incurred_cost(source)
  process_part_payment
end

def process_part_payment
  return process_evidence_check if source.is_a?(PartPayment)
  fee(source)
end

def process_evidence_check
  another_amount = if source.application.evidence_check.present?
                     source.application.evidence_check
                   else
                     source.application
                   end
  work_out_the_sum(another_amount)
end

def work_out_the_sum(another_amount)
  fee(@source) - amount_to_pay(another_amount)
end
```

or maybe even:

```ruby

def incurred_cost
  process_part_payment
end

def process_part_payment
  return work_out_the_sum if @source.is_a?(PartPayment)
  fee(source)
end

def process_evidence_check
  if @source.application.evidence_check.present?
    @source.application.evidence_check
  else
    @source.application
  end
end

def work_out_the_sum
  fee(@source) - amount_to_pay(process_evidence_check)
end
```